### ISSUE TITLE
kafka-cscout does not produce call graphs

### ISSUE THREAD

**Issue by MagielBruntink** [fasten-project/debian-builder#5]
Dear @gdrosos and @StefanosChaliasos,

I'm having an issue with cscout at the moment - it seems it cannot generate callgraphs anymore. I tried running as a Kafka plugin like in the monster deployment - no luck. Also the debug example from the README does not work.

To replicate:
```
docker run --rm --privileged ghcr.io/fasten-project/debian-builder/kafka-cscout:1.0.4 --directory /callgraphs --debug "{'source': 'anna', 'version': '1.71', 'arch': 'amd64', 'release': 'buster',  'date': ''}"
```

Output I get:
```
2021-12-18 19:43:05.738105: Phase: begin Sent:  to log_topic
2021-12-18 19:43:08.926699: Run sbuild for anna-buster-amd64-1.71, dist=buster, arch=amd64
2021-12-18 19:43:09.971666: File not found: /callgraphs/anna/buster/1.71/amd64/report
2021-12-18 19:43:09.971697: Push error message to kafka topic: report: File not found: /callgraphs/anna/buster/1.71/amd64/report
2021-12-18 19:43:09.971719: Phase: error Sent:  to error_topic
2021-12-18 19:43:09.971936: Phase: failed Sent:  to log_topic
2021-12-18 19:43:09.972020: Cleanup
2021-12-18 19:43:09.972210: /callgraphs/anna/buster/1.71/amd64/ - No such file or directory.
```
Could you please help me out? The FASTEN pipeline sure gets better with C callgraphs :-)

**Comment by StefanosChaliasos** [comment]
TL;DR: Problem solved; there is still an issue when processing buster packages, for which I propose two solutions in the following.

The underlying error was that buster Is not still the stable version of Debian. When I tried the `sbuild-cscout` image, I got the following error.

```
docker run -it --rm --privileged -v $(pwd)/callgraphs:/callgraphs \
    schaliasos/sbuild-cscout sbuild --apt-update --no-apt-upgrade \
    --no-apt-distupgrade --batch --stats-dir=/var/log/sbuild/stats \
    --dist=buster --arch=amd64 zlib
Use of uninitialized value in concatenation (.) or string at /usr/bin/sbuild line 329.
Use of uninitialized value in concatenation (.) or string at /usr/bin/sbuild line 329.
sbuild (Debian sbuild) 0.78.1 (09 February 2019) on d25ef634590b

+==============================================================================+
| zlib (amd64)                                 Mon, 20 Dec 2021 17:02:19 +0000 |
+==============================================================================+

Package: zlib
Distribution: buster
Machine Architecture: amd64
Host Architecture: amd64
Build Architecture: amd64
Build Type: binary

I: NOTICE: Log filtering will replace 'var/run/schroot/mount/buster-amd64-sbuild-a1dd257e-0445-4ebd-98a4-f06653364104' with '<<CHROOT>>'
I: NOTICE: Log filtering will replace 'build/zlib-5ubbGL/resolver-txLtOj' with '<<RESOLVERDIR>>'

+------------------------------------------------------------------------------+
| Update chroot                                                                |
+------------------------------------------------------------------------------+

Get:1 http://deb.debian.org/debian buster InRelease [122 kB]
Reading package lists...
E: Repository 'http://deb.debian.org/debian buster InRelease' changed its 'Suite' value from 'stable' to 'oldstable'
E: apt-get update failed
Use of uninitialized value in concatenation (.) or string at /usr/bin/sbuild line 329.

+------------------------------------------------------------------------------+
| Summary                                                                      |
+------------------------------------------------------------------------------+

Build Architecture: amd64
Build Type: binary
Build-Space: 0
Build-Time: 0
Distribution: buster
Fail-Stage: apt-get-update
Host Architecture: amd64
Install-Time: 0
Job: zlib
Machine Architecture: amd64
Package: zlib
Package-Time: 0
Space: 0
Status: failed
--------------------------------------------------------------------------------
Finished at 2021-12-20T17:02:19Z
Build needed 00:00:00, 0k disk space
Use of uninitialized value in concatenation (.) or string at /usr/bin/sbuild line 329.
Use of uninitialized value in concatenation (.) or string at /usr/bin/sbuild line 329.
E: apt-get update failed
```

A [minor fix](https://github.com/fasten-project/debian-builder/commit/914c3c50b94275e96607156ea418c216badc7884), and updating the images fixed the issue. 

```
docker run -it --privileged     -v $(pwd)/temp/debug:/home/builder/debug     -v $(pwd)/temp/debian:/mnt/fasten/debian     kafka-cscout --directory /mnt/fasten/debian     --debug "{'source': 'anna', 'version': '1.83', 'arch': 'amd64', 'release': 'bullseye',  'date': ''}"
2021-12-20 20:59:18.368737: Phase: begin Sent:  to log_topic
2021-12-20 20:59:22.580783: Run sbuild for anna-bullseye-amd64-1.83, dist=bullseye, arch=amd64
2021-12-20 21:08:42.574146: Push call graph for anna to kafka topic
2021-12-20 21:08:42.576837: Phase: succeed Sent:  to produce_topic
2021-12-20 21:08:42.577037: Call graph generated
2021-12-20 21:08:42.577072: Phase: complete Sent:  to log_topic
2021-12-20 21:08:42.577217: Cleanup
```

Yet, it fails to produce the call graphs when analyzing buster packages.
The following errors occur when running CScout.

```
cscout: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by cscout)
cscout: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by cscout)
```

The issue here is that we build CScout in the Docker image, and we use that version in the sbuild. Previously, both of them have Buster (and the same version of libraries), but now the Docker image has Bullseye, and sbuild has Buster. 

There are two solutions:

1. To build different Docker images for every Debian release you want to support.
2. To build CScout inside sbuild each time you analyze a package.

**Comment by StefanosChaliasos** [comment]
@gdrosos Please update the docker images and publish them under the fasten namespace.

**Comment by MagielBruntink** [comment]
Awesome @StefanosChaliasos thank you!

**Comment by gdrosos** [comment]
I uploaded and published separately all the docker images of cscout for the production of call graphs for buster (old stable) & bullseye (new stable) release packages, in order to be employed independently.

The new docker image regarding generating call graphs for buster packages is published under the namespace: 
``` docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-buster:latest```

To adduce an example using the image:
```
docker run -it --privileged     -v $(pwd)/temp/debug:/home/builder/debug     -v $(pwd)/temp/debian:/mnt/fasten/debian     docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-buster:latest  --directory /mnt/fasten/debian     --debug "{'source': 'anna', 'version': '1.71', 'arch': 'amd64', 'release': 'buster',  'date': ''}"
...
Status: Downloaded newer image for docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-buster:latest
2021-12-23 17:13:24.540117: Phase: begin Sent:  to log_topic
2021-12-23 17:13:28.376088: Run sbuild for anna-buster-amd64-1.71, dist=buster, arch=amd64
2021-12-23 17:14:37.141822: Push call graph for anna to kafka topic
2021-12-23 17:14:37.145821: Phase: succeed Sent:  to produce_topic
2021-12-23 17:14:37.147166: Call graph generated
2021-12-23 17:14:37.147219: Phase: complete Sent:  to log_topic
2021-12-23 17:14:37.147508: Cleanup
```

The docker image regarding generating call graphs for packages of the bullseye debian release is published under the namespace: ``` docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-bullseye:latest```.

To adduce an example using the image:
```
docker run -it --privileged     -v $(pwd)/temp/debug:/home/builder/debug     -v $(pwd)/temp/debian:/mnt/fasten/debian     docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-bullseye:latest --directory /mnt/fasten/debian     --debug "{'source': 'anna', 'version': '1.83', 'arch': 'amd64', 'release': 'bullseye',  'date': ''}"
...
Status: Downloaded newer image for docker.pkg.github.com/fasten-project/debian-builder/kafka-cscout-bullseye:latest
2021-12-23 22:02:01.133211: Phase: begin Sent:  to log_topic
2021-12-23 22:02:05.275936: Run sbuild for anna-bullseye-amd64-1.83, dist=bullseye, arch=amd64
2021-12-23 22:03:31.016891: Push call graph for anna to kafka topic
2021-12-23 22:03:31.020579: Phase: succeed Sent:  to produce_topic
2021-12-23 22:03:31.020906: Call graph generated
2021-12-23 22:03:31.020979: Phase: complete Sent:  to log_topic
2021-12-23 22:03:31.021164: Cleanup
```

**Comment by MagielBruntink** [comment]
I can confirm this issue has been fixed, thanks guys!
